lib/Sema/CSBindings.cpp

```diff
@@ -751,3 +751,101 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
 
   return result;
 }
+
+/// \brief Enumerates all of the 'direct' supertypes of the given type.
+///
+/// The direct supertype S of a type T is a supertype of T (e.g., T < S)
+/// such that there is no type U where T < U and U < S.
+static SmallVector<Type, 4> enumerateDirectSupertypes(TypeChecker &tc,
+                                                      Type type) {
+  SmallVector<Type, 4> result;
+
+  if (type->mayHaveSuperclass()) {
+    // FIXME: Can also weaken to the set of protocol constraints, but only
+    // if there are any protocols that the type conforms to but the superclass
+    // does not.
+
+    // If there is a superclass, it is a direct supertype.
+    if (auto superclass = tc.getSuperClassOf(type))
+      result.push_back(superclass);
+  }
+
+  if (!type->isMaterializable())
```

Comment:
Can this just be `type->is<InOutType>() || type->is<LValueType>()`? Or maybe it can only be one of those two?

Reply:
Maybe, once again it's just me moving code around, I'll take a deeper look at what is going on inside :)